lib/babe/build.go

```diff
@@ -237,6 +237,147 @@ func (b *BlockBuilder) buildBlockExtrinsics(slot Slot, rt runtime.Instance) []*t
 	return included
 }
 
+// CandidateDescriptor is a unique descriptor of the candidate receipt.
+type CandidateDescriptor struct {
+	// The ID of the para this is a candidate for.
+	ParaID uint32
+	// The hash of the relay-chain block this should be executed in
+	// the context of.
+	// NOTE: the fact that the hash includes this value means that code depends
+	// on this for deduplication. Removing this field is likely to break things.
+	RelayParent common.Hash
+	// The collator's relay-chain account ID
+	Collator []byte // CollatorId
+	// Signature on blake2-256 of components of this receipt:
+	// The para ID, the relay parent, and the `pov_hash`.
+	// TODO: I have made a guess that all signatures are []byte
+	Signature []byte // CollatorSignature
+	// The hash of the `pov-block`.
+	PovHash common.Hash
+}
+
+// UpwardMessage is a message from a parachain to its Relay Chain.
+type UpwardMessage []byte
+
+// OutboundHrmpMessage is an HRMP message seen from the perspective of a sender.
+type OutboundHrmpMessage struct {
+	Recipient uint32
+	Data      []byte
+}
+
+// All Vec<u8> in rust have become []byte here
+
+// ValidationCode is Parachain validation code.
+type ValidationCode []byte
+
+// HeadData is Parachain head data included in the chain.
+type HeadData []byte
```

Comment:
Do we want all these types if they don't have any methods defined on them?

Maybe just using `[]byte` directly would be clearer, with the right variable names. That would also reduce the Go API which is already too large imo.

Reply:
in substrate you mean right?

Reply:
Nope, polkadot. That's where all these structs are. https://github.com/paritytech/polkadot/blob/v0.9.17/primitives/src/v1/mod.rs#L1345

Reply:
Interesting!

@timwu20 what do you think we should do in cases like such (I'd like to add it to the code style document of the repo): should we mirror what the Polkadot Rust code has or not care about it?

Reply:
Yes I think we should.  Even though they are just byte arrays right now, I'd like to know what they are.  We can add methods down the road if we need to.  It would be nice to be able to search through both codebases and see something similar.  Ideally I'd like to break this out into a separate repo the way parity does it.